src/histogram.c

```diff
@@ -0,0 +1,236 @@
+ * if they are below or above the range, respectively. Array representations are one-based by 
+ * default in PostgreSQL whereas they are zero-based in C, hence the histogram function operates 
+ * on index conversion with extra variables (i.e. s) in the function code. The function assumes
+ * a zero index in the elems array regardless of the bucket count. It is converted to a zero- or 
+ * one-based array in PostgreSQL depending on the bucket count. If the histogram array contains
+ * a zero index, it will be noted by the outputted histogram indices (e.g. [0:2]={5,1,0}).
+ */
+
+PG_FUNCTION_INFO_V1(hist_sfunc);
+PG_FUNCTION_INFO_V1(hist_combinefunc);
+
+/* Generate a histogram */
+Datum
+hist_sfunc(PG_FUNCTION_ARGS) 
+{
+	MemoryContext aggcontext; 
+	ArrayType 	*state = PG_ARGISNULL(0) ? NULL : PG_GETARG_ARRAYTYPE_P(0);
+	Datum 		*elems; 
+
+	float 	val = PG_GETARG_FLOAT4(1); 
+	float 	min = PG_GETARG_FLOAT4(2); 
+	float 	max = PG_GETARG_FLOAT4(3); 
+	int 	nbuckets = PG_GETARG_INT32(4); 
+
+	int 	bucket = DirectFunctionCall4(width_bucket_float8, val, min, max, nbuckets);  
+
+	int     dims[1]; // 1-D array containing number of buckets used to construct histogram
+ 	int     lbs[1]; // 1-D array containing the lower bound used to construct histogram
+ 	int 	s = 0; // Extra indexing variable for conversion between C and PostgreSQL arrays
+
+ 	/* Determine the lower bound (i.e. zero- or one-basing in PostgreSQL array) */
+ 	lbs[0] = (bucket == 0) ? 0 : 1;
+
+	if (!AggCheckCallContext(fcinfo, &aggcontext))
+	{
+		/* Cannot be called directly because of internal-type argument */
+		elog(ERROR, "hist_sfunc called in non-aggregate context");
+	}
+
+
+	if (min > max) {
+		elog(ERROR, "lower bound cannot exceed upper bound");
+	}
+
+	/* Init the array with the correct number of 0's so the caller doesn't see NULLs (for loop) */
+	if (state == NULL)  
+	{
+		/* Increment the number of buckets if width_bucket returned nbuckets + 1 */
+		if (bucket > nbuckets) {
+			nbuckets++;
+		}
+
+		/* Construct a zero-based array representation in C (+1 accounts for the zero-th element) */
+		elems = (Datum *) MemoryContextAlloc(aggcontext, sizeof(Datum) * (nbuckets + 1)); 
+
+		for (int i = 0; i <= nbuckets; i++) {
+			elems[i] = (Datum) 0;
+		}
+
+		/* Specify the number of elements by accounting for buckets outside of the input range */
+		dims[0] = nbuckets + 1 - lbs[0]; 
+	}
+
+	else { 
+		Oid    	i_eltype;
+	    int16  	i_typlen;
+	    bool   	i_typbyval;
+	    char   	i_typalign;
+	    int 	n;
+	    bool 	*nulls;
+	    /* Copy of elems if needed */
+	    Datum 	*elems_edit;
+
+		/* Get input array element type */
+		i_eltype = ARR_ELEMTYPE(state);
+		get_typlenbyvalalign(i_eltype, &i_typlen, &i_typbyval, &i_typalign);
+
+		/* Deconstruct array into a default zero-based elems array*/
+		deconstruct_array(state, i_eltype, i_typlen, i_typbyval, i_typalign, &elems, &nulls, &n); 
```

Comment:
Is deconstructing preferred as opposed to just using the Array API, such as `array_set_element()` and `array_get_element()`?